src/mapgen.cpp

```diff
@@ -3544,19 +3544,23 @@ C..C..C...|hhh|#########\n\
                 }
                 if (t_above == "lab_stairs" || t_above == "ice_lab_stairs") {
                     int sx, sy;
+                    int attempts = 100;
                     do {
                         sx = rng(lw, SEEX * 2 - 1 - rw);
                         sy = rng(tw, SEEY * 2 - 1 - bw);
-                    } while (ter(sx, sy) != t_rock_floor);
+                        attempts--;
+                    } while ( ( ter( sx, sy ) != t_rock_floor ) && attempts && !g->m.has_furn( sx,sy ) );
                     ter_set(sx, sy, t_stairs_up);
                 }
 
                 if (terrain_type == "lab_stairs" || terrain_type == "ice_lab_stairs") {
                     int sx, sy;
+                    int attempts = 100;
                     do {
                         sx = rng(lw, SEEX * 2 - 1 - rw);
                         sy = rng(tw, SEEY * 2 - 1 - bw);
-                    } while (ter(sx, sy) != t_rock_floor);
+                        attempts--;
+                    } while ( ( ter( sx, sy ) != t_rock_floor ) && attempts && !g->m.has_furn( sx,sy ) );
                     ter_set(sx, sy, t_stairs_down);
```

Comment:
These two blocks are almost identical (except `t_stairs_*`). Please, consider extracting a method. The whole `map::draw_map` is a complete mess already and we have to start clearing it up someday. Also, the most proper (and the fastest in case of small amount of suitable spots) way of putting something at random is building a vector of suitable spots and picking an item. It doesn't depend on max number of attempts and has O(n) complexity. The current implementation can put a stair on `t_rock_floor` after 100 attempts even if a better place exists.